hazardous/metrics/_concordance_index.py

```diff
@@ -0,0 +1,471 @@
+        tied_tol=tied_tol,
+    )
+    return np.array(c_index_report["cindex"])
+
+
+def _concordance_index_incidence_report(
+    y_test,
+    y_pred,
+    y_train=None,
+    ipcw_estimator="km",
+    time_grid=None,
+    taus=None,
+    event_of_interest=1,
+    tied_tol=1e-8,
+):
+    """Report version of function ``concordance_index_incidence``.
+
+    Running this function directly is useful to get more insights about
+    the underlying statistics of the C-index.
+    All returned lists have the same length as taus.
+
+    Returns
+    -------
+    cindex: list of float
+        Value of the concordance index
+
+    n_pairs_a: list of int
+        Number of comparable pairs with (T_i < T_j) | ((T_i = T_j) & (D_j = 0))
+        (type A).
+        Those are the only comparable pairs for survival without competing events.
+
+    n_concordant_pairs_a: list of int
+        Number of concordant pairs among A pairs without ties.
+
+    n_ties_times: list of int
+        Number of pairs which experienced the event of interest at the same time
+        (D_i = D_j = 1) & (T_i = T_j).
+
+    n_ties_pred_a: list of int
+        Number of pairs of type A with np.abs(y_pred_i - y_pred_j) <= tied_tol
+
+    n_pairs_b: list of int
+        Number of comparable pairs with T_i >= T_j where j has
+        a competing event (type B).
+        0 if there are no competing events.
+
+    n_concordant_pairs_b: list of int
+        Number of concordant pairs among B pairs without ties.
+
+    n_ties_pred_b: list of int
+        Number of pairs of type B with np.abs(y_pred_i - y_pred_j) <= tied_tol.
+    """
+    if y_pred.ndim != 2:
+        raise ValueError(
+            "y_pred dimension must be 2, with shape (n_samples, n_time_grid), "
+            f"got {y_pred.ndim=}."
+        )
+    if y_test.shape[0] != y_pred.shape[0]:
+        raise ValueError(
+            "y_test and y_pred must be the same length, got:"
+            f"{y_test.shape[0]=} and {y_pred.shape[0]=}."
+        )
+    if time_grid is not None and len(time_grid) != y_pred.shape[1]:
+        raise ValueError(
+            "y_pred must have the same number of columns as the length of time_grid, "
+            f"got: {y_test.shape[1]=} and {len(time_grid)=}."
+        )
+
+    if ipcw_estimator is None:
+        if y_train is not None:
+            warnings.warn(
+                "y_train passed but ipcw_estimator is set to None, "
+                "therefore y_train won't be used. Set y_train=None "
+                "to silence this warning."
+            )
+        ipcw = np.ones(y_test["event"].shape[0])
+
+    else:
+        if y_train is None:
+            # Raising here since the error raised by the IPCW estimator doesn't
+            # help the user understand that y_train is missing.
+            raise ValueError(
+                "ipcw_estimator is set, but y_train is None. "
+                "Set y_train to fix this error."
+            )
+        # TODO: add cox option
```

Comment:
maybe explicit why cox would be relevant here (just say that it would allow the censoring model to be conditional)?